src/backend/links.js

```diff
@@ -77,32 +73,128 @@ const cacheLink = fetchPolicy =>
   });
 
 export const initLinkEvents = (hook, bridge) => {
-  // handle incoming requests
+  // Handle incoming requests
   const subscriber = request => {
     const { query, variables, operationName, key, fetchPolicy } = JSON.parse(
       request,
     );
+
     try {
-      const userLink = hook.ApolloClient.link;
-      const cache = hook.ApolloClient.cache;
-
-      const devtoolsLink = from([
-        errorLink(),
-        cacheLink(fetchPolicy),
-        schemaLink(),
-        userLink,
-      ]);
-      const obs = execute(devtoolsLink, {
-        query: gql(query),
-        variables,
-        operationName,
-        context: { __devtools_key__: key, cache },
-      });
-      obs.subscribe({
-        next: data => bridge.send(`link:next:${key}`, JSON.stringify(data)),
-        error: err => bridge.send(`link:error:${key}`, JSON.stringify(err)),
+      const apolloClient = hook.ApolloClient;
+      const userLink = apolloClient.link;
+      const cache = apolloClient.cache;
+      let operationExecution$;
+      const queryAst = gql(query);
+      const subscriptionHandlers = {
+        next(data) {
+          bridge.send(`link:next:${key}`, JSON.stringify(data));
+        },
+        error(err) {
+          bridge.send(`link:error:${key}`, JSON.stringify(err));
+        },
         complete: () => bridge.send(`link:complete:${key}`),
+      };
+
+      // Devtools can currently be used with 2 versions of local state
+      // handling: 1) Using `apollo-link-state` or 2) Using local state
+      // features integrated directly into Apollo Client. `apollo-link-state`
+      // will eventually be deprecated, but for the time being we need to
+      // support both approaches via devtools.
+      //
+      // The `apollo-link-state` approach uses a custom link chain to parse
+      // and execute queries, whereas the Apollo Client local state approach
+      // uses Apollo Client directly. To decide which approach to use
+      // below, we'll check to see if typeDefs have been set on the
+      // ApolloClient instance, as if so, this means Apollo Client local state
+      // is being used.
+
+      const supportsApolloClientLocalState =
+        typeof apolloClient.getTypeDefs === "function";
```

Comment:
@hwillson Just a note that https://github.com/apollographql/apollo-client/pull/4361/commits/d3931b8347c76c961ae24da68ef108576e67ff0d may have broken this logic (which is fine as long as we make the logic work again before we ship `apollo-client@2.5.0`).

Reply:
With this commit, the new way to get the client `typeDefs` will be `apolloClient.typeDefs`: https://github.com/apollographql/apollo-client/pull/4361/commits/adfd9c008c1a9afab09fdda70f0b121287ed6eb7

Reply:
Thank you @benjamn ! @hwillson had warned me that `getTypeDefs` was going to be replaced so I've been ready for the breaking change 😃 If this will be released in the latest alpha I'll update the DevTools PR and merge.